pkg/operator/controller/configurable-route/controller.go

```diff
@@ -0,0 +1,402 @@
+	}
+
+	reconciler := &reconciler{
+		kclient:       kubeClient,
+		config:        config,
+		client:        mgr.GetClient(),
+		cache:         mgr.GetCache(),
+		eventRecorder: eventRecorder,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Trigger reconcile requests for the cluster ingress resource.
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+		clusterIngressResource := operatorcontroller.IngressClusterConfigName()
+		return meta.GetName() == clusterIngressResource.Name && meta.GetNamespace() == clusterIngressResource.Namespace
+	})
+
+	if err := c.Watch(&source.Kind{Type: &configv1.Ingress{}}, &handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+
+	// Trigger reconcile requests for the roles and roleBindings with the componentRoute label.
+	defaultPredicate := predicate.NewPredicateFuncs(func(meta metav1.Object, object runtime.Object) bool {
+		labels := meta.GetLabels()
+		_, ok := labels[componentRouteHashAnnotation]
+		return ok
+	})
+
+	roleInformer, err := mgr.GetCache().GetInformer(context.TODO(), &rbacv1.Role{})
+	if err := c.Watch(&source.Informer{Informer: roleInformer}, &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(reconciler.resourceToClusterIngressConfig)}, defaultPredicate); err != nil {
+		return nil, err
+	}
+
+	roleBindingInformer, err := mgr.GetCache().GetInformer(context.TODO(), &rbacv1.RoleBinding{})
+	if err := c.Watch(&source.Informer{Informer: roleBindingInformer}, &handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(reconciler.resourceToClusterIngressConfig)}, defaultPredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// resourceToClusterIngressConfig is used to only trigger reconciles on the cluster ingress config
+func (r *reconciler) resourceToClusterIngressConfig(o handler.MapObject) []reconcile.Request {
+	return []reconcile.Request{
+		{
+			operatorcontroller.IngressClusterConfigName(),
+		},
+	}
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	SecretNamespace string
+}
+
+// reconciler handles the actual ingress reconciliation logic in response to
+// events.
+type reconciler struct {
+	config        Config
+	client        client.Client
+	kclient       kubernetes.Interface
+	cache         cache.Cache
+	eventRecorder events.Recorder
+}
+
+// Reconcile expects request to refer to a ingress in the operator namespace,
+// and will do all the work to ensure the ingress is in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Only proceed if we can get the ingress resource.
+	ingress := &configv1.Ingress{}
+	if err := r.cache.Get(context.TODO(), request.NamespacedName, ingress); err != nil {
```

Comment:
why do we care about ingresses? Aren't the route names in the config virtual, i.e. without 1:1 match with real routes?

Reply:
The `ingresses.config.openshift.io/cluster` resource icontains the route configuration information. The CRD was updated in [this PR](https://github.com/openshift/api/pull/852).